xds/src/main/java/io/grpc/xds/sds/SdsSslContextProvider.java

```diff
@@ -0,0 +1,269 @@
+        updateSslContext();
+      }
+    } else if (secretUpdate.hasValidationContext()) {
+      checkState(
+          secretUpdate.getName().equals(validationContextSdsConfig.getName()),
+          "validationContext names don't match");
+      certificateValidationContext = secretUpdate.getValidationContext();
+      if (tlsCertificate != null || certSdsConfig == null) {
+        updateSslContext();
+      }
+    } else {
+      throw new UnsupportedOperationException(
+          "Unexpected secret type:" + secretUpdate.getTypeCase());
+    }
+  }
+
+  // this gets called only when requested secrets are ready...
+  private void updateSslContext() {
+    try {
+      SslContextBuilder sslContextBuilder;
+      if (server) {
+        sslContextBuilder =
+            GrpcSslContexts.forServer(
+                tlsCertificate.getCertificateChain().getInlineBytes().newInput(),
+                tlsCertificate.getPrivateKey().getInlineBytes().newInput(),
+                tlsCertificate.hasPassword()
+                    ? tlsCertificate.getPassword().getInlineString()
+                    : null);
+        if (certificateValidationContext != null) {
+          sslContextBuilder.trustManager(new SdsTrustManagerFactory(certificateValidationContext));
+        }
+      } else {
+        sslContextBuilder =
+            GrpcSslContexts.forClient()
+                .trustManager(new SdsTrustManagerFactory(certificateValidationContext));
+        if (tlsCertificate != null) {
+          sslContextBuilder.keyManager(
+              tlsCertificate.getCertificateChain().getInlineBytes().newInput(),
+              tlsCertificate.getPrivateKey().getInlineBytes().newInput(),
+              tlsCertificate.hasPassword() ? tlsCertificate.getPassword().getInlineString() : null);
+        }
+      }
+      SslContext sslContextCopy = sslContextBuilder.build();
+      sslContext = sslContextCopy;
+      makePendingCallbacks(sslContextCopy);
+    } catch (CertificateException | IOException | CertStoreException e) {
+      logger.log(Level.SEVERE, "", e);
+    }
+  }
+
+  private void makePendingCallbacks(SslContext sslContextCopy) {
+    synchronized (pendingCallbacks) {
+      for (Pair pair : pendingCallbacks) {
+        callPerformCallback(pair.callback, pair.executor, sslContextCopy);
+      }
+      pendingCallbacks.clear();
+    }
+  }
+
+  @Override
+  public void onError(Status error) {
+    synchronized (pendingCallbacks) {
+      for (Pair pair : pendingCallbacks) {
+        pair.callback.onException(error.asException());
+      }
+      pendingCallbacks.clear();
+    }
+  }
+
+  @Override
+  void close() {
+    if (certSdsClient != null) {
+      certSdsClient.cancelSecretWatch(this);
+      certSdsClient.shutdown();
+    }
+    if (validationContextSdsClient != null) {
+      validationContextSdsClient.cancelSecretWatch(this);
+      validationContextSdsClient.shutdown();
+    }
+  }
+
+  private static class Pair {
```

Comment:
Pair exists in guava and in general has special meaning (and it is not recommended to use). this is very specific type of pair that can have a better name.